mdf_matio/validator.py

```diff
@@ -0,0 +1,397 @@
+
+        # Require strict JSON
+        try:
+            json.dumps(ds_md, allow_nan=False)
+        except (ValueError, json.JSONDecodeError) as e:
+            return {
+                "success": False,
+                "error": "Invalid dataset JSON: {}".format(str(e)),
+                "details": repr(e)
+            }
+
+        # Remove null/None values
+        ds_md = _remove_nulls(ds_md, self.__allowed_nulls)
+
+        # Validate against schema
+        try:
+            jsonschema.validate(ds_md, schema, resolver=resolver)
+        except jsonschema.ValidationError as e:
+            return {
+                "success": False,
+                "error": "Invalid dataset metadata: " + str(e).split("\n")[0],
+                "details": str(e)
+            }
+
+        # Check projects blocks allowed
+        # If no blocks, disallow projects
+        if not self.__project_blocks:
+            if ds_md.get("projects"):
+                return {
+                    "success": False,
+                    "error": "Unauthorized project metadata: No projects allowed",
+                    "details": "'project' block not allowed: '{}'".format(ds_md)
+                }
+        # If some project blocks allowed, check that only allowed ones are present
+        else:
+            unauthorized = []
+            for proj in ds_md.get("projects", {}).keys():
+                if proj not in self.__project_blocks:
+                    unauthorized.append(proj)
+            if unauthorized:
+                return {
+                    "success": False,
+                    "error": ("Unauthorized project metadata: '{}' not allowed"
+                              .format(unauthorized)),
+                    "details": ("Not authorized for project block(s) '{}' in '{}'. "
+                                "The dataset is not in an allowed organization."
+                                .format(unauthorized, ds_md))
+                }
+
+        # Validate required fields
+        # TODO: How should this validation be done?
+        # The metadata conforms to the schema, there are just extra
+        # `requires` values. Perhaps add these to the schema instead?
+        # Lists, specifically, are an issue. Must all dicts in the list
+        # conform? This behavior is difficult.
+        # As a semi-temporary measure, only check the first element of lists.
+        if self.__required_fields:
+            missing = []
+            for field_path in self.__required_fields:
+                value = ds_md
+                for field_name in field_path.split("."):
+                    try:
+                        value = value[field_name]
+                        if isinstance(value, list) and len(value) > 0:
+                            value = value[0]
+                    except KeyError:
+                        missing.append(field_path)
+                        break
+            if missing:
+                return {
```

Comment:
It should be OK to throw exceptions instead, given how you have "if not success then raise exception" to catch these "exceptions." Not using exceptions masks that the error actually came from here in a stacktrace.